compiler/rustc_parse/src/lexer/tokentrees.rs

```diff
@@ -48,220 +48,210 @@ impl<'a> TokenTreesReader<'a> {
         let mut buf = TokenStreamBuilder::default();
 
         self.bump();
-        while self.token != token::Eof {
-            buf.push(self.parse_token_tree()?);
+        loop {
+            match self.token.kind {
+                token::OpenDelim(delim) => buf.push(self.parse_token_tree_open_delim(delim)),
+                token::CloseDelim(delim) => return Err(self.close_delim_err(delim)),
+                token::Eof => return Ok(buf.into_token_stream()),
+                _ => buf.push(self.parse_token_tree_other()),
+            }
         }
-
-        Ok(buf.into_token_stream())
     }
 
     // Parse a stream of tokens into a list of `TokenTree`s, up to a `CloseDelim`.
     fn parse_token_trees_until_close_delim(&mut self) -> TokenStream {
         let mut buf = TokenStreamBuilder::default();
         loop {
-            if let token::CloseDelim(..) = self.token.kind {
-                return buf.into_token_stream();
-            }
-
-            match self.parse_token_tree() {
-                Ok(tree) => buf.push(tree),
-                Err(mut e) => {
-                    e.emit();
+            match self.token.kind {
+                token::OpenDelim(delim) => buf.push(self.parse_token_tree_open_delim(delim)),
+                token::CloseDelim(..) => return buf.into_token_stream(),
```

Comment:
`parse_all_token_trees` and `parse_token_trees_until_close_delim` are currently very similar, and this would make them less similar, because it would change the latter but not the former. I'd rather keep the similarity in place. (I tried merging them at one point but it ended up making things a little slower, but it's possible that merging them in the future would make sense.)